From c7de99ecbf7adb689b08e1d5c8958ef48c5851d5 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Mon, 12 Jun 2023 15:31:54 +0200
Subject: [PATCH 061/120] watchdog: arm_smc_wdt: get wdt status through
 SMCWD_GET_TIMELEFT

The optional SMCWD_GET_TIMELEFT command can be used to detect if
the watchdog has already been started.

Run wdt_start() on already started watchdog to keep going pinging
it, even when property 'u-boot,noautostart' is present.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 97a964002aec00cc7577de104cbf8a2ff4fb3417 watchdog: arm_smc_wdt: get wdt status through SMCWD_GET_TIMELEFT]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I4d8e2aec1b1b47a9fbbbe7a25004c644f26906a9
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/313802
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/watchdog/arm_smc_wdt.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/watchdog/arm_smc_wdt.c b/drivers/watchdog/arm_smc_wdt.c
index 0ea44445700..ee8a41fd478 100644
--- a/drivers/watchdog/arm_smc_wdt.c
+++ b/drivers/watchdog/arm_smc_wdt.c
@@ -46,6 +46,8 @@ static int smcwd_call(struct udevice *dev, enum smcwd_call call,
 		return -ENODEV;
 	if (res->a0 == PSCI_RET_INVALID_PARAMS)
 		return -EINVAL;
+	if (res->a0 == PSCI_RET_DISABLED)
+		return -ENODATA;
 	if (res->a0 != PSCI_RET_SUCCESS)
 		return -EIO;
 
@@ -99,6 +101,21 @@ static int smcwd_probe(struct udevice *dev)
 	priv->min_timeout = res.a1;
 	priv->max_timeout = res.a2;
 
+	/* If already started, then force u-boot to use it */
+	err = smcwd_call(dev, SMCWD_GET_TIMELEFT, 0, NULL);
+	switch (err) {
+	case 0:
+		dev_dbg(dev, "Already started\n");
+		wdt_set_force_start(dev);
+		break;
+	case -ENODATA:
+		dev_dbg(dev, "Not already started\n");
+		break;
+	default:
+		/* Optional SMCWD_GET_TIMELEFT not implemented */
+		break;
+	}
+
 	return 0;
 }
 
-- 
2.47.2

